app/src/main/java/io/apicurio/registry/storage/impl/sql/AbstractSqlRegistryStorage.java

```diff
@@ -731,6 +731,17 @@ protected ArtifactMetaDataDto createArtifactWithMetadata(String groupId, String
         String createdBy = securityIdentity.getPrincipal().getName();
         Date createdOn = new Date();
 
+        if (groupId != null && !isGroupExists(groupId)) {
+            //Only create group metadata for non-default groups.
+            createGroup(GroupMetaDataDto.builder()
+                    .groupId(groupId)
+                    .createdOn(0)
+                    .modifiedOn(0)
+                    .createdBy(createdBy)
+                    .modifiedBy(createdBy)
+                    .build());
+        }
```

Comment:
Note that we probably have a race condition here.  Group existence will be checked and if it doesn't exist it will be created.  But it might be created by another request in the meantime.  In this case, we don't really care.  So in the case of the race condition, the create request will fail with a "group already exists exception".  Which is going to be a weird response from the server.

I suggest we have a variant of createGroup (or a param to the existing method) that will ignore the PK violation rather than throw the exception.  We can keep the `isGroupExists()` check here since it seems like a reasonable quick check.  But then we want to **not fail** if the group actually does exist when we go to create it.

Or maybe it's more complicated than this?  What if the group is created by another user in the meantime and owner-only auth (with group access) is enabled?  The race condition might allow a user to create an artifact in a group they didn't create and don't own.

WDYT?  Perhaps we just leave this as-is.  The edge case/race condition will fail with a strange response back to the user, but that might be better than trying to figure out what to do.

Reply:
Actually, I realized this and I gave it some thought. I think it's actually better to just fail since trying to handle this scenario opens the door to all sorts of edge cases in regard to authorization, just as you described. We would need to check for group existence outside of the storage so that we can then apply authorization rules, maybe then undo changes in the database depending on some conditions etc. In this particular case, I think leaving it as-is is probably the best.

Reply:
Yeah great - I think I talked myself into the same conclusion as I typed up the review note.  :)

Thanks.